### PR TITLE
Rollup Guide

### PR DESCRIPTION
**Pull request by AdityaSripal**
Writeup descibing how to implement a rollup client.

Since the architectures of different rollups are wildly different, I cannot write an ICS standard. Instead I've written english explanations of what to do for each function, things to consider, and some pseudocode that makes liberal use of unspecified functions that would be different for each rollup implementation but have documentation and naming that makes it clear what is expected of the functions.

Open to suggestions on where this code should live.

### CODE REVIEW

File: spec/client/Rollup-Integration-guide.md [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+```typescript
+function updateState(clientMessage: ClientMessage) {
+    // marshalling logic omitted
+    header = clientMessage.header
+    consensusState = ConsensusState(header.appHash, header.time)
+
+    setConsensusState(clientMessage.clientId, clientMessage.height, consensusState)
+
+    // create mapping between consensus state and the current time for fraud proof waiting period
+    setSubmitTime(consensusState, currentTime())
```
Comment (crodriguezvega) [inline spec/client/Rollup-Integration-guide.md]: In ICS 07 we do the following:

```suggestion
    provableStore.set("clients/{clientMessage.clientId}/processedTimes/{header.GetHeight()}", currentTimestamp())
```

File: spec/client/Rollup-Integration-guide.md [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    // check conditional clients are still valid
+    rollupClient = getClient(clientState.DALayer)
+    settlementClient = getClient(clientState.settlementLayer) // may not exist for all rollups
+
+    assert(isActive(clientState))
+    assert(isActive(rollupClient))
+    assert(isActive(settlementClient))
+
+    consensusState = getConsensusState(clientState.clientId, height)
+    submitTime = getSubmitTime(consensusState)
```
Comment (crodriguezvega) [inline spec/client/Rollup-Integration-guide.md:169]: In ICS 07 we call this `processedTime`:

```suggestion
    processedTime = provableStore.set("clients/{clientState.clientId}/processedTimes/{height}")
```

File: spec/client/Rollup-Integration-guide.md [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+): error {
+    // check conditional clients are still valid
+    rollupClient = getClient(clientState.DALayer)
+    settlementClient = getClient(clientState.settlementLayer) // may not exist for all rollups
+
+    assert(isActive(clientState))
+    assert(isActive(rollupClient))
+    assert(isActive(settlementClient))
+
+    consensusState = getConsensusState(clientState.clientId, height)
```
Comment (crodriguezvega) [inline spec/client/Rollup-Integration-guide.md:168]:
```suggestion
    consensusState = provableStore.get("clients/{clientState.clientId}/consensusStates/{height}")
```

File: spec/client/Rollup-Integration-guide.md [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
```
Comment (crodriguezvega) [inline spec/client/Rollup-Integration-guide.md:193]:
```suggestion
    consensusState = provableStore.get("clients/{clientState.clientId}/consensusStates/{height}")
```

File: spec/client/Rollup-Integration-guide.md [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    submitTime = getSubmitTime(consensusState)
+
+    // must ensure fraud proof period has passed
+    assert(submitTime + clientState.fraudPeriod > currentTime())
```
Comment (crodriguezvega) [inline spec/client/Rollup-Integration-guide.md]: Is `fraudPeriod` then something we should store in the `ClientState` of the Rollkit client?

File: spec/client/Rollup-Integration-guide.md [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+    assert(proof.verifyMembership(path, value))
```
Comment (crodriguezvega) [inline spec/client/Rollup-Integration-guide.md:174]: In ICS 07 we do it like this:

```suggestion
    if !verifyMembership(consensusState.commitmentRoot, proof, path, value) {
        return error
    }
    return nil
```

File: spec/client/Rollup-Integration-guide.md [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+    assert(isActive(rollupClient))
+    assert(isActive(settlementClient))
+
+    consensusState = getConsensusState(clientState.clientId, height)
+    submitTime = getSubmitTime(consensusState)
+
+    // must ensure fraud proof period has passed
+    assert(submitTime + clientState.fraudPeriod > currentTime())
+
+    assert(proof.verifyNonMembership(path))
```
Comment (crodriguezvega) [inline spec/client/Rollup-Integration-guide.md:199]:
```suggestion
    if !verifyNonMembership(consensusState.commitmentRoot, proof, path) {
        return error
    }
    return nil
```

File: spec/client/Rollup-Integration-guide.md [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+
+The updateState for rollups works the same as typical clients, though it is critical that the optimistic rollup client stores the submit time for when the consensus state was created so that we can verify that the fraud period has passed.
+
+```typescript
+function updateState(clientMessage: ClientMessage) {
+    // marshalling logic omitted
+    header = clientMessage.header
+    consensusState = ConsensusState(header.appHash, header.time)
+
+    setConsensusState(clientMessage.clientId, clientMessage.height, consensusState)
```
Comment (crodriguezvega) [inline spec/client/Rollup-Integration-guide.md:68]:
```suggestion
    provableStore.set("clients/{clientMessage.clientId}/consensusStates/{header.GetHeight()}", consensusState)
```

File: spec/client/Rollup-Integration-guide.md [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+
+The optimistic fraud proof verifier should be implemented as a smart contract. Since the fraud prover is dependent not on the consensus but on the application state machine itself. Thus each rollup instance needs its own fraud prover. Having each fraud prover encoded directly in the client requires a different implementation for each rollup instance. Instead, calling out to a separate smart contract allows the client to be reused for all instances, and for new fraud provers to be uploaded for a new rollup application.
+
+```typescript
+// optimistic rollup fraud proof
+// the misbehaviour must be associated with a height on the rollup
+function checkForMisbehaviour(clientMessage: ClientMessage) {
+    // unmarshalling logic ommitted
+    clientId = clientMessage.clientId
+    misbehaviour = clientMessage.misbehaviour
```
Comment (crodriguezvega) [inline spec/client/Rollup-Integration-guide.md:95]: Just to confirm: there are no checks we need to perform on `clientMessage.header` for misbehaviour detection then?

File: spec/client/Rollup-Integration-guide.md [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+
+ZK rollups can verify correctness of the header upon submission since the rollup client can embed a proving circuit that can verify a ZK proof from the relayer that the submitted header is correct. Optimistic rollups on the other hand cannot immediately trust a header upon submission, as the header may later be proved fraudulent. Thus, the header can be stored but must wait for the fraud period to elapse without any successful challenges to the correctness of the header before it is finalized and used for proof verification.
+
+```typescript
+function verifyClientMessage(clientMessage: ClientMessage) {
+    clientState = getClient(clientMessage.clientId)
+    // note: unmarshalling logic omitted
+    // verify the header against the rollups own consensus mechanism if it exists
+    // e.g. verify sequencer signature
+    verifySignatures(clientMessage.header, clientSequencers)
```
Comment (crodriguezvega) [inline spec/client/Rollup-Integration-guide.md:21]: Do we need to do any verifications is `clientMessage` is actually of `Misbehaviour` type?

File: spec/client/Rollup-Integration-guide.md [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+    // is rollup-specific so it is left as an unspecified function
+    // in this document.
+    daClient = getClient(clientState.DALayer)
+    verifyMembership(
+        daClient,
+        clientMessage.DAProofHeight,
+        0,
+        0,
+        clientMessage.DAHeaderProof,
+        DAHeaderPath(clientState.clientId, clientMessage.height),
```
Comment (crodriguezvega) [inline spec/client/Rollup-Integration-guide.md:35]: This should be the client ID of the DA client, right?

```suggestion
        DAHeaderPath(daClient.clientId, clientMessage.height),
```

File: spec/client/Rollup-Integration-guide.md [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+        clientMessage.header)
+    verifyMembership(
+        daClient,
+        clientMessage.DAProofHeight,
+        0,
+        0,
+        clientMessage.DABlockDataProof,
+        DABlockDataPath(clientState.clientId, clientMessage.height),
```
Comment (crodriguezvega) [inline spec/client/Rollup-Integration-guide.md:43]: Same here I think:

```suggestion
        DABlockDataPath(daClient.clientId, clientMessage.height),
```

**Review by crodriguezvega — COMMENTED** [review]
Thanks, @AdityaSripal. Left for now a bunch of suggestions to keep consistency with the other specs.

File: spec/client/Rollup-Integration-guide.md (hunk 10 above)
Comment (AdityaSripal) [reply spec/client/Rollup-Integration-guide.md:32]: Fixed

File: spec/client/Rollup-Integration-guide.md [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+
+ZK rollups can verify correctness of the header upon submission since the rollup client can embed a proving circuit that can verify a ZK proof from the relayer that the submitted header is correct. Optimistic rollups on the other hand cannot immediately trust a header upon submission, as the header may later be proved fraudulent. Thus, the header can be stored but must wait for the fraud period to elapse without any successful challenges to the correctness of the header before it is finalized and used for proof verification.
+
+```typescript
+function verifyClientMessage(clientMessage: ClientMessage) {
+    switch typeof(clientMsg) {
```
Comment (sangier) [inline spec/client/Rollup-Integration-guide.md:17]:
```suggestion
    switch typeof(clientMessage) {
```

File: spec/client/Rollup-Integration-guide.md [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+    case Header:
+      verifyHeader(clientMsg)
```
Comment (sangier) [inline spec/client/Rollup-Integration-guide.md:19]:
```suggestion
      verifyHeader(clientMessage)
```

File: spec/client/Rollup-Integration-guide.md [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+    assert(isActive(rollupClient))
+    assert(isActive(settlementClient))
+
+    consensusState = provableStore.get("clients/{clientState.clientId}/consensusStates/{height}")
+    processedTime = provableStore.set("clients/{clientState.clientId}/processedTimes/{height}")
+
+    // must ensure fraud proof period has passed
+    assert(submitTime + clientState.fraudPeriod > currentTimestamp())
+
+    if !verifyMembership(consensusState.commitmentRoot, proof, path, value) {
```
Comment (sangier) [inline spec/client/Rollup-Integration-guide.md:187]: Should this call include the clientState parameter? 
```suggestion

    if !verifyMembership(clientState, consensusState.commitmentRoot, proof, path, value) {
```

File: spec/client/Rollup-Integration-guide.md [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+    assert(isActive(rollupClient))
+    assert(isActive(settlementClient))
+
+    consensusState = provableStore.get("clients/{clientState.clientId}/consensusStates/{height}")
+    submitTime = getSubmitTime(consensusState)
+
+    // must ensure fraud proof period has passed
+    assert(submitTime + clientState.fraudPeriod > currentTimestamp())
+
+    if !verifyNonMembership(consensusState.commitmentRoot, proof, path) {
```
Comment (sangier) [inline spec/client/Rollup-Integration-guide.md:215]: Same here: 
```suggestion
    if !verifyNonMembership(clientState, consensusState.commitmentRoot, proof, path) {
```

Comment (sangier) [reply spec/client/Rollup-Integration-guide.md:215]: No, in Ics007 we do the same here.

**Review by sangier — APPROVED** [review]
Thanks @AdityaSripal for this work, looks as a really good starting point! I think is good to keep it in the client folder for now. And thanks @crodriguezvega for the review!

File: spec/client/Rollup-Integration-guide.md (hunk 5 above)
Comment (AdityaSripal) [reply spec/client/Rollup-Integration-guide.md]: Yes, we could theoretically put it in the connection and encode as the delay time. But I think that is more confusing.

We should discuss with rollup teams as to what makes more sense

File: spec/client/Rollup-Integration-guide.md [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+  path: CommitmentPath,
+  value: bytes
+): Error {
+  // check conditional clients are still valid
+  rollupClient = getClient(clientState.DALayer)
+  settlementClient = getClient(clientState.settlementLayer) // may not exist for all rollups
+
+  assert(isActive(clientState))
+  assert(isActive(rollupClient))
+  assert(isActive(settlementClient))
```
Comment (damiannolan) [inline spec/client/Rollup-Integration-guide.md:180]: The `rollupClient` here is the `clientState` passed as an arg. i.e the execution client?

```suggestion
  daClient = getClient(clientState.DALayer)
  settlementClient = getClient(clientState.settlementLayer) // may not exist for all rollups

  assert(isActive(clientState))
  assert(isActive(daClient))
  assert(isActive(settlementClient))
```

File: spec/client/Rollup-Integration-guide.md [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+  proof: CommitmentProof,
+  path: CommitmentPath,
+): Error {
+  // check conditional clients are still valid
+  rollupClient = getClient(clientState.DALayer)
+  settlementClient = getClient(clientState.settlementLayer) // may not exist for all rollups
+
+  assert(isActive(clientState))
+  assert(isActive(rollupClient))
+  assert(isActive(settlementClient))
```
Comment (damiannolan) [inline spec/client/Rollup-Integration-guide.md:208]: ditto 

```suggestion
  daClient = getClient(clientState.DALayer)
  settlementClient = getClient(clientState.settlementLayer) // may not exist for all rollups

  assert(isActive(clientState))
  assert(isActive(daClient))
  assert(isActive(settlementClient))
```

**Review by damiannolan — APPROVED** [review]
Excellent work on this, really nice explanations. A little bit is unclear for me around the settlement client if/else branch I highlighted in a comment, but maybe we can discuss that soon. 

I was thinking to also suggest that it may be nice to add a short glossary of terms at the top of page and/or prerequisite readings. Some short and concise definitions would be nice, will just jot down some that come to mind here (can always take this in a follow up PR too):
- Data Availability layer
- Execution layer, and sequencer
- Settlement layer
- Different client types for above. e.g. Rollup/execution client etc
- ZK

File: spec/client/Rollup-Integration-guide.md [hunk 19]
```diff
@@ -0,0 +1,221 @@
...
+// optimistic rollup fraud proof
+// the misbehaviour must be associated with a height on the rollup
+function checkForMisbehaviour(clientMessage: ClientMessage) {
+  // unmarshalling logic ommitted
+  misbehaviour = Misbehaviour(clientMessage)
+  clientId = clientMessage.clientId
+  client = getClient(clientId)
+  // if the rollup has a settlement layer, we can delegate the fraud proof game to the settlement layer
+  // and simply verify with the settlement client that fraud has been proven for the given misbehaviour
+  if client.settlementLayer == nil {
```
[the remark this answers is not in the document]
Comment (AdityaSripal) [reply spec/client/Rollup-Integration-guide.md]: If the rollup has a settlement layer, it means that there is a blockchain whose entire job is to verify fraud proofs of that rollup.

So we don't need to fraud prove on our own chain, we can just rely on the settlement client to do that work just as off-chain rollup light clients will do.

If the fraud window passes without a fraud proof submitted to the settlement layer, we can assume the block is valid

**Pull request merged**